File: le3/src/le3_light.cpp
#include "le3_light.h"

#include <fmt/core.h>

const char* gLightSpriteSuffix = "_sprite";

LE3AmbientLight::LE3AmbientLight(glm::vec3 color, float intensity) : 
    LE3Object("ambient_light"), color(color), intensity(intensity), m_pRigidBody(nullptr)
{
}
void LE3AmbientLight::Update(double deltaTime)
{
    LE3Object::Update(deltaTime);

    if (m_pRigidBody)
    {
        btTransform transform;
        transform.setFromOpenGLMatrix(glm::value_ptr(GetModelMatrix() * glm::scale(glm::vec3(1/m_scale))));
        m_pRigidBody->setCenterOfMassTransform(transform);
    }
}
void LE3AmbientLight::Draw()
{
    if (GetSprite())
        GetSprite()->GetMaterial()->diffuseColor = glm::vec4(GetColor(), 1.f);
}
glm::vec3 LE3AmbientLight::GetColor() const
{
    return this->color;
}
void LE3AmbientLight::SetColor(glm::vec3 color)
{
    this->color = color;
}

float LE3AmbientLight::GetIntensity() const
{
    return this->intensity;
}
void LE3AmbientLight::SetIntensity(float intensity)
{
    this->intensity = intensity;
}
LE3Sprite* LE3AmbientLight::GetSprite()
{
    for (auto child : m_children)
        if (dynamic_cast<LE3Sprite*>(child) && 
            (child->GetName() == GetName() + gLightSpriteSuffix))
            return dynamic_cast<LE3Sprite*>(child);
    
    return nullptr;
}
void LE3AmbientLight::SetName(std::string name)
{
    LE3Object::SetName(name);

    LE3Sprite* sprite = GetSprite();
    if (sprite)
        sprite->SetName(GetName() + gLightSpriteSuffix);
}


LE3DirectionalLight::LE3DirectionalLight(std::string name, glm::vec3 color, float intensity) : 
    LE3Object(name), color(color), intensity(intensity), bEnableShadows(false), m_pRigidBody(nullptr), bDebugLine(false)
{
}
void LE3DirectionalLight::Update(double deltaTime)
{
    LE3Object::Update(deltaTime);

    if (m_pRigidBody)
    {
        btTransform transform;
        transform.setFromOpenGLMatrix(glm::value_ptr(GetModelMatrix() * glm::scale(glm::vec3(1/m_scale))));
        m_pRigidBody->setCenterOfMassTransform(transform);
    }
}
void LE3DirectionalLight::Draw()
{
    if (GetSprite())
        GetSprite()->GetMaterial()->diffuseColor = glm::vec4(GetColor(), 1.f);

    if (bDebugLine)
        LE3VisualDebug::DrawDebugLine(
            GetGlobalPosition(), 
            GetGlobalPosition() + gLightDebugRayLength * GetDirection(), 
            GetColor());
}
glm::vec3 LE3DirectionalLight::GetColor() const
{
    return this->color;
}
void LE3DirectionalLight::SetColor(glm::vec3 color)
{
    this->color = color;
}

float LE3DirectionalLight::GetIntensity() const
{
    return this->intensity;
}
void LE3DirectionalLight::SetIntensity(float intensity)
{
    this->intensity = intensity;
}
glm::vec3 LE3DirectionalLight::GetDirection() const
{
    glm::vec3 dir = this->GetModelMatrix() * glm::vec4(gDefaultLightDirection, 0.f);
    return dir;
}
glm::mat4 LE3DirectionalLight::GetViewMatrix(glm::vec3 pos) const
{
    float delta_plane = 10.f;
    glm::mat4 lightProjection = glm::ortho(
        pos.x - delta_plane, pos.x + delta_plane, 
        pos.y - delta_plane, pos.y + delta_plane, 
        pos.z - 2.f, pos.z + 100.f);
    // To combat linearly dependant columns in look-at matrix, we add a very small noise to the up vector
    // glm::mat4 lightView = glm::lookAt(pos, pos + GetDirection(), glm::vec3(0.f, .999f, 0.04471017781f));
    glm::mat4 lightView = glm::lookAt(-delta_plane * GetDirection(), glm::vec3(0.f), glm::vec3(0.f, .999f, 0.04471017781f));
    return lightProjection * lightView;
}
bool LE3DirectionalLight::IsShadowsEnabled() const
{
    return bEnableShadows;
}
void LE3DirectionalLight::SetShadowsEnabled(bool enabled)
{
    bEnableShadows = enabled;
}
LE3ShadowMap& LE3DirectionalLight::GetShadowMap()
{
    return shadowMap;
}
LE3Sprite* LE3DirectionalLight::GetSprite()
{
    for (auto child : m_children)
        if (dynamic_cast<LE3Sprite*>(child) && 
            (child->GetName() == GetName() + gLightSpriteSuffix))
            return dynamic_cast<LE3Sprite*>(child);
    
    return nullptr;
}
void LE3DirectionalLight::SetName(std::string name)
{
    LE3Object::SetName(name);

    LE3Sprite* sprite = GetSprite();
    if (sprite)
        sprite->SetName(GetName() + gLightSpriteSuffix);
}


LE3PointLight::LE3PointLight(std::string name, glm::vec3 color, float intensity, float attn_const, float attn_linear, float attn_exp) : 
    LE3Object(name), color(color), intensity(intensity), attn_const(attn_const), attn_linear(attn_linear), attn_exp(attn_exp), m_pRigidBody(nullptr)
{
}
void LE3PointLight::Update(double deltaTime)
{
    LE3Object::Update(deltaTime);

    if (m_pRigidBody)
    {
        btTransform transform;
        transform.setFromOpenGLMatrix(glm::value_ptr(GetModelMatrix() * glm::scale(glm::vec3(1/m_scale))));
        m_pRigidBody->setCenterOfMassTransform(transform);
    }
}
void LE3PointLight::Draw()
{
    if (GetSprite())
        GetSprite()->GetMaterial()->diffuseColor = glm::vec4(GetColor(), 1.f);
}
glm::vec3 LE3PointLight::GetColor() const
{
    return this->color;
}
void LE3PointLight::SetColor(glm::vec3 color)
{
    this->color = color;
}

float LE3PointLight::GetIntensity() const
{
    return this->intensity;
}
void LE3PointLight::SetIntensity(float intensity)
{
    this->intensity = intensity;
}
float LE3PointLight::GetAttenuationConst() const
{
    return this->attn_const;
}
float LE3PointLight::GetAttenuationLinear() const
{
    return this->attn_linear;
}
float LE3PointLight::GetAttenuationExp() const
{
    return this->attn_exp;
}
void LE3PointLight::SetAttenuationConst(float attn_const)
{
    this->attn_const = attn_const;
}
void LE3PointLight::SetAttenuationLinear(float attn_linear)
{
    this->attn_linear = attn_linear;
}
void LE3PointLight::SetAttenuationExp(float attn_exp)
{
    this->attn_exp = attn_exp;
}
LE3Sprite* LE3PointLight::GetSprite()
{
    for (auto child : m_children)
        if (dynamic_cast<LE3Sprite*>(child) && 
            (child->GetName() == GetName() + gLightSpriteSuffix))
            return dynamic_cast<LE3Sprite*>(child);
    
    return nullptr;
}
void LE3PointLight::SetName(std::string name)
{
    LE3Object::SetName(name);

    LE3Sprite* sprite = GetSprite();
    if (sprite)
        sprite->SetName(GetName() + gLightSpriteSuffix);
}

LE3SpotLight::LE3SpotLight(std::string name, glm::vec3 color, float intensity, 
        float cutoff, float outer_cutoff) :
    LE3Object(name), color(color), intensity(intensity), cutoff(cutoff), outer_cutoff(outer_cutoff), 
    bEnableShadows(false), m_pRigidBody(nullptr), bDebugLine(false)
{
}

void LE3SpotLight::Update(double deltaTime)
{
    LE3Object::Update(deltaTime);

    if (m_pRigidBody)
    {
        btTransform transform;
        transform.setFromOpenGLMatrix(glm::value_ptr(GetModelMatrix() * glm::scale(glm::vec3(1/m_scale))));
        m_pRigidBody->setCenterOfMassTransform(transform);
    }
}
void LE3SpotLight::Draw()
{
    if (GetSprite())
        GetSprite()->GetMaterial()->diffuseColor = glm::vec4(GetColor(), 1.f);
    if (bDebugLine)
        LE3VisualDebug::DrawDebugLine(
            GetGlobalPosition(), 
            GetGlobalPosition() + gLightDebugRayLength * GetDirection(), 
            GetColor());
}
glm::vec3 LE3SpotLight::GetColor() const
{
    return this->color;
}
void LE3SpotLight::SetColor(glm::vec3 color)
{
    this->color = color;
}

float LE3SpotLight::GetIntensity() const
{
    return this->intensity;
}
void LE3SpotLight::SetIntensity(float intensity)
{
    this->intensity = intensity;
}

float LE3SpotLight::GetCutoff() const
{
    return this->cutoff;
}
float LE3SpotLight::GetOuterCutoff() const
{
    return this->outer_cutoff;
}
void LE3SpotLight::SetCutoff(float cutoff)
{
    this->cutoff = cutoff;
}
void LE3SpotLight::SetOuterCutoff(float outer_cutoff)
{
    this->outer_cutoff = cutoff;
}

glm::vec3 LE3SpotLight::GetDirection() const
{
    glm::vec3 dir = this->GetModelMatrix() * glm::vec4(gDefaultLightDirection, 0.f);
    return dir;
}
glm::mat4 LE3SpotLight::GetViewMatrix() const
{
    float delta_plane = 15.f;
    glm::vec3 pos = GetGlobalPosition();
    glm::mat4 lightProjection = glm::perspective(glm::radians(45.f), 1.f, .5f, 40.f);
    // To combat linearly dependant columns in look-at matrix, we add a very small noise to the up vector
    glm::mat4 lightView = glm::lookAt(pos, pos + GetDirection(), glm::vec3(0.f, .999f, 0.04471017781f));
    return lightProjection * lightView;
}
bool LE3SpotLight::IsShadowsEnabled() const
{
    return bEnableShadows;
}
void LE3SpotLight::SetShadowsEnabled(bool enabled)
{
    bEnableShadows = enabled;
}
LE3ShadowMap& LE3SpotLight::GetShadowMap()
{
    return shadowMap;
}
LE3Sprite* LE3SpotLight::GetSprite()
{
    for (auto child : m_children)
        if (dynamic_cast<LE3Sprite*>(child) && 
            (child->GetName() == GetName() + gLightSpriteSuffix))
            return dynamic_cast<LE3Sprite*>(child);
    
    return nullptr;
}
void LE3SpotLight::SetName(std::string name)
{
    LE3Object::SetName(name);

    LE3Sprite* sprite = GetSprite();
    if (sprite)
        sprite->SetName(GetName() + gLightSpriteSuffix);
}

File: resources/engine/scripts/bindings/objects/LE3Light.lua
LE3Light.__base = LE3Object
LE3Light.load = function (scene, tbl, res)
    assert(res ~= nil)
    LE3Light.__base.load(scene, tbl, res)

    if (tbl.Color ~= nil) then LE3Light.set_color(res, table.unpack(tbl.Color)) end
    if (tbl.Intensity ~= nil) then LE3Light.set_intensity(res, tbl.Intensity) end
    return {ptr = res, name = tbl.Name}
end
LE3Light.rebuild = function (object, tbl)
    object = LE3Light.__base.rebuild(object, tbl)
    if (tbl.Color ~= nil) then LE3Light.set_color(object, table.unpack(tbl.Color)) end
    if (tbl.Intensity ~= nil) then LE3Light.set_intensity(object, tbl.Intensity) end
    return object
end
LE3Light.save = function (object)
    local tbl = LE3Light.__base.save(object)
    tbl.Color = {LE3Light.get_color(object)}
    tbl.Intensity = LE3Light.get_intensity(object)
    return tbl
end
LE3Light.title = "LE3Light"
LE3Light.properties = {
    {name = "Color", type = "color"},
    {name = "Intensity", type = "float"}
}


LE3AmbientLight.__base = LE3Light
LE3AmbientLight.load = function (scene, tbl, res)
    if (res == nil) then 
        LE3Scene.add_ambient_light(scene, tbl.Name)
        res = LE3Scene.get_object(scene, tbl.Name)
    end
    LE3AmbientLight.__base.load(scene, tbl, res)
    return {ptr = res, name = tbl.Name}
end
LE3AmbientLight.rebuild = function (object, tbl)
    LE3AmbientLight.__base.rebuild(object, tbl)
end
LE3AmbientLight.save = function (object)
    local tbl = LE3AmbientLight.__base.save(object)
    return tbl
end
LE3AmbientLight.title = "LE3AmbientLight"


LE3DirectionalLight.__base = LE3Light
LE3DirectionalLight.load = function (scene, tbl, res)
    if (res == nil) then
        LE3Scene.add_directional_light(scene, tbl.Name)
        res = LE3Scene.get_object(scene, tbl.Name)
    end
    LE3DirectionalLight.__base.load(scene, tbl, res)

    if (tbl.HasShadowMap) then LE3DirectionalLight.add_shadow_map(res, LE3EngineConfig.Rendering.ShadowMapResolution) end
    return {ptr = res, name = tbl.Name}
end
LE3DirectionalLight.rebuild = function (object, tbl)
    LE3DirectionalLight.__base.rebuild(object, tbl)
end
LE3DirectionalLight.save = function (object)
    local tbl = LE3DirectionalLight.__base.save(object)
    tbl.HasShadowMap = LE3DirectionalLight.has_shadow_map(object)
    return tbl
end
LE3DirectionalLight.title = "LE3DirectionalLight"


LE3PointLight.__base = LE3Light
LE3PointLight.load = function (scene, tbl, res)
    if (res == nil) then 
        LE3Scene.add_point_light(scene, tbl.Name)
        res = LE3Scene.get_object(scene, tbl.Name)
    end
    LE3PointLight.__base.load(scene, tbl, res)

    if (tbl.AttnConst ~= nil) then LE3PointLight.set_attn_const(res, tbl.AttnConst) end
    if (tbl.AttnLinear ~= nil) then LE3PointLight.set_attn_linear(res, tbl.AttnLinear) end
    if (tbl.AttnExp ~= nil) then LE3PointLight.set_attn_exp(res, tbl.AttnExp) end
    return {ptr = res, name = tbl.Name}
end
LE3PointLight.rebuild = function (object, tbl)
    LE3PointLight.__base.rebuild(object, tbl)
    if (tbl.AttnConst ~= nil) then LE3PointLight.set_attn_const(object, tbl.AttnConst) end
    if (tbl.AttnLinear ~= nil) then LE3PointLight.set_attn_linear(object, tbl.AttnLinear) end
    if (tbl.AttnExp ~= nil) then LE3PointLight.set_attn_exp(object, tbl.AttnExp) end
end
LE3PointLight.save = function (object)
    local tbl = LE3PointLight.__base.save(object)
    tbl.AttnConst = LE3PointLight.get_attn_const(object)
    tbl.AttnLinear = LE3PointLight.get_attn_linear(object)
    tbl.AttnExp = LE3PointLight.get_attn_exp(object)
    return tbl
end
LE3PointLight.title = "LE3PointLight"
LE3PointLight.properties = {
    {name = "AttnConst", type = "float"},
    {name = "AttnLinear", type = "float"},
    {name = "AttnExp", type = "float"}
}


LE3SpotLight.__base = LE3Light
LE3SpotLight.load = function (scene, tbl, res)
    if (res == nil) then 
        LE3Scene.add_spot_light(scene, tbl.Name)
        res = LE3Scene.get_object(scene, tbl.Name)
    end
    LE3SpotLight.__base.load(scene, tbl, res)

    if (tbl.Cutoff ~= nil) then LE3SpotLight.set_cutoff(res, tbl.Cutoff) end
    if (tbl.OuterCutoff ~= nil) then LE3SpotLight.set_outer_cutoff(res, tbl.OuterCutoff) end

    if (tbl.HasShadowMap) then LE3SpotLight.add_shadow_map(res, LE3EngineConfig.Rendering.ShadowMapResolution) end
    return {ptr = res, name = tbl.Name}
end
LE3SpotLight.rebuild = function (object, tbl)
    LE3SpotLight.__base.rebuild(object, tbl)
    if (tbl.Cutoff ~= nil) then LE3SpotLight.set_cutoff(object, tbl.Cutoff) end
    if (tbl.OuterCutoff ~= nil) then LE3SpotLight.set_outer_cutoff(object, tbl.OuterCutoff) end
end
LE3SpotLight.save = function (object)
    local tbl = LE3SpotLight.__base.save(object)
    tbl.Cutoff = LE3SpotLight.get_cutoff(object)
    tbl.OuterCutoff = LE3SpotLight.get_outer_cutoff(object)
    tbl.HasShadowMap = LE3SpotLight.has_shadow_map(object)
    return tbl
end
LE3SpotLight.title = "LE3SpotLight"
LE3SpotLight.properties = {
    {name = "Cutoff", type = "float"},
    {name = "OuterCutoff", type = "float"}
}